File: backend/src/main/java/com/tailorstudio/app/payment/PhonePeClient.java
package com.tailorstudio.app.payment;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * PhonePe Payment Gateway Standard Checkout (OAuth + create payment + order status).
 * Docs: https://developer.phonepe.com/payment-gateway/website-integration/standard-checkout/
 */
@Component
public class PhonePeClient {

    private final PhonePeProperties props;
    private final ObjectMapper objectMapper;
    private final HttpClient http =
            HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(20)).build();

    private volatile String cachedBearer;
    private volatile Instant tokenGoodUntil = Instant.EPOCH;

    public PhonePeClient(PhonePeProperties props, ObjectMapper objectMapper) {
        this.props = props;
        this.objectMapper = objectMapper;
    }

    public boolean isReady() {
        return props.isEnabled() && props.hasCredentials();
    }

    public synchronized String getAuthorizationHeader() throws IOException, InterruptedException {
        if (cachedBearer != null && Instant.now().isBefore(tokenGoodUntil.minusSeconds(60))) {
            return "O-Bearer " + cachedBearer;
        }
        String form =
                Stream.of(
                                "client_id=" + url(props.getClientId()),
                                "client_version=" + url(props.getClientVersion()),
                                "client_secret=" + url(props.getClientSecret()),
                                "grant_type=" + url("client_credentials"))
                        .collect(Collectors.joining("&"));
        HttpRequest req =
                HttpRequest.newBuilder()
                        .uri(URI.create(props.oauthTokenUrl()))
                        .timeout(Duration.ofSeconds(25))
                        .header("Content-Type", "application/x-www-form-urlencoded")
                        .POST(HttpRequest.BodyPublishers.ofString(form))
                        .build();
        HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2) {
            throw new IOException("PhonePe token HTTP " + res.statusCode() + ": " + res.body());
        }
        JsonNode root = objectMapper.readTree(res.body());
        cachedBearer = root.path("access_token").asText(null);
        if (cachedBearer == null || cachedBearer.isEmpty()) {
            throw new IOException("PhonePe token response missing access_token");
        }
        long exp = root.path("expires_at").asLong(0);
        if (exp > 0) {
            tokenGoodUntil = Instant.ofEpochSecond(exp);
        } else {
            tokenGoodUntil = Instant.now().plusSeconds(25 * 60);
        }
        return "O-Bearer " + cachedBearer;
    }

    public PhonePePayResult createCheckout(long amountPaisa, String merchantOrderId, String redirectUrl)
            throws IOException, InterruptedException {
        if (amountPaisa < 100) {
            throw new IllegalArgumentException("PhonePe minimum amount is 100 paise (₹1)");
        }
        String auth = getAuthorizationHeader();
        ObjectNode body = objectMapper.createObjectNode();
        body.put("merchantOrderId", merchantOrderId);
        body.put("amount", amountPaisa);
        body.put("expireAfter", 1800);
        ObjectNode flow = body.putObject("paymentFlow");
        flow.put("type", "PG_CHECKOUT");
        flow.put("message", "Tailor order payment");
        flow.putObject("merchantUrls").put("redirectUrl", redirectUrl);
        ObjectNode meta = body.putObject("metaInfo");
        meta.put("udf1", merchantOrderId);

        HttpRequest.Builder b =
                HttpRequest.newBuilder()
                        .uri(URI.create(props.checkoutPayUrl()))
                        .timeout(Duration.ofSeconds(30))
                        .header("Content-Type", "application/json")
                        .header("Authorization", auth)
                        .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(body)));
        if (props.getMerchantId() != null && !props.getMerchantId().isBlank()) {
            b.header("X-MERCHANT-ID", props.getMerchantId());
        }
        HttpResponse<String> res = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2) {
            throw new IOException("PhonePe create payment HTTP " + res.statusCode() + ": " + res.body());
        }
        JsonNode root = objectMapper.readTree(res.body());
        String redirect = root.path("redirectUrl").asText(null);
        if (redirect == null || redirect.isEmpty()) {
            throw new IOException("PhonePe response missing redirectUrl: " + res.body());
        }
        String orderId = root.path("orderId").asText("");
        return new PhonePePayResult(redirect, orderId);
    }

    public JsonNode getOrderStatus(String merchantOrderId) throws IOException, InterruptedException {
        String auth = getAuthorizationHeader();
        HttpRequest.Builder b =
                HttpRequest.newBuilder()
                        .uri(URI.create(props.orderStatusUrl(merchantOrderId)))
                        .timeout(Duration.ofSeconds(25))
                        .header("Content-Type", "application/json")
                        .header("Authorization", auth)
                        .GET();
        if (props.getMerchantId() != null && !props.getMerchantId().isBlank()) {
            b.header("X-MERCHANT-ID", props.getMerchantId());
        }
        HttpResponse<String> res = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2) {
            throw new IOException("PhonePe status HTTP " + res.statusCode() + ": " + res.body());
        }
        return objectMapper.readTree(res.body());
    }

    private static String url(String s) {
        return URLEncoder.encode(s == null ? "" : s, StandardCharsets.UTF_8);
    }

    public record PhonePePayResult(String redirectUrl, String phonePeOrderId) {}
}

File: server/src/payment/phonePeClient.ts
import { env } from '../config/env.js'

let cachedBearer: string | null = null
let tokenUntil = 0

const enc = (s: string) => encodeURIComponent(s)

export function phonePeReady() {
  return env.phonepe.enabled && env.phonepe.clientId && env.phonepe.clientSecret
}

function oauthUrl() {
  return env.phonepe.sandbox
    ? 'https://api-preprod.phonepe.com/apis/pg-sandbox/v1/oauth/token'
    : 'https://api.phonepe.com/apis/identity-manager/v1/oauth/token'
}

function payUrl() {
  return env.phonepe.sandbox
    ? 'https://api-preprod.phonepe.com/apis/pg-sandbox/checkout/v2/pay'
    : 'https://api.phonepe.com/apis/pg/checkout/v2/pay'
}

function statusUrl(merchantOrderId: string) {
  const base = env.phonepe.sandbox
    ? 'https://api-preprod.phonepe.com/apis/pg-sandbox/checkout/v2/order/'
    : 'https://api.phonepe.com/apis/pg/checkout/v2/order/'
  return `${base}${enc(merchantOrderId)}/status?details=false&errorContext=false`
}

async function authHeader(): Promise<string> {
  if (cachedBearer && Date.now() < tokenUntil - 60_000) return `O-Bearer ${cachedBearer}`
  const body = new URLSearchParams({
    client_id: env.phonepe.clientId,
    client_version: env.phonepe.clientVersion,
    client_secret: env.phonepe.clientSecret,
    grant_type: 'client_credentials',
  })
  const res = await fetch(oauthUrl(), {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: body.toString(),
  })
  if (!res.ok) throw new Error(`PhonePe token HTTP ${res.status}`)
  const j = (await res.json()) as { access_token?: string; expires_at?: number }
  if (!j.access_token) throw new Error('PhonePe token missing access_token')
  cachedBearer = j.access_token
  tokenUntil = j.expires_at ? j.expires_at * 1000 : Date.now() + 25 * 60_000
  return `O-Bearer ${cachedBearer}`
}

export async function createCheckout(amountPaisa: number, merchantOrderId: string, redirectUrl: string) {
  const auth = await authHeader()
  const headers: Record<string, string> = { 'Content-Type': 'application/json', Authorization: auth }
  if (env.phonepe.merchantId) headers['X-MERCHANT-ID'] = env.phonepe.merchantId
  const res = await fetch(payUrl(), {
    method: 'POST',
    headers,
    body: JSON.stringify({
      merchantOrderId,
      amount: amountPaisa,
      expireAfter: 1800,
      paymentFlow: { type: 'PG_CHECKOUT', message: 'Tailor order payment', merchantUrls: { redirectUrl } },
      metaInfo: { udf1: merchantOrderId },
    }),
  })
  if (!res.ok) throw new Error(`PhonePe create HTTP ${res.status}: ${await res.text()}`)
  const j = (await res.json()) as { redirectUrl?: string; orderId?: string }
  if (!j.redirectUrl) throw new Error('PhonePe missing redirectUrl')
  return { redirectUrl: j.redirectUrl, phonePeOrderId: j.orderId || '' }
}

export async function getOrderStatus(merchantOrderId: string) {
  const auth = await authHeader()
  const headers: Record<string, string> = { 'Content-Type': 'application/json', Authorization: auth }
  if (env.phonepe.merchantId) headers['X-MERCHANT-ID'] = env.phonepe.merchantId
  const res = await fetch(statusUrl(merchantOrderId), { headers })
  if (!res.ok) throw new Error(`PhonePe status HTTP ${res.status}`)
  return res.json() as Promise<{ state?: string }>
}
